specs/phase1/beacon-chain.md

```diff
@@ -767,236 +755,175 @@ def validate_attestation(state: BeaconState, attestation: Attestation) -> None:
+            aggregation_bits=attestation.aggregation_bits
+        ))
 ```
 
 ##### Shard transition processing
 
-###### `apply_shard_transition`
+###### `execute_shard_transition`
 
 ```python
-def apply_shard_transition(state: BeaconState, shard: Shard, transition: ShardTransition) -> None:
-    # TODO: only need to check it once when phase 1 starts
-    assert state.slot > PHASE_1_FORK_SLOT
-
-    # Correct data root count
-    offset_slots = get_offset_slots(state, shard)
+def execute_shard_transition(state: BeaconState, shard: Shard, transition: ShardTransition) -> ShardState:
+    """
+    Applies the given ShardTransition to the appropriate ShardState. Also processes shard proposer rewards and fees.
+    """
+    shard_state = state.shard_states[shard]
+    prev_gasprice = shard_state.gasprice
+    shard_parent_root = shard_state.latest_block_root
+    admissible_slots = compute_admissible_slots(transition.start_slot)[:len(transition.shard_data_roots)]
     assert (
         len(transition.shard_data_roots)
         == len(transition.shard_states)
         == len(transition.shard_block_lengths)
-        == len(offset_slots)
+        == len(admissible_slots)
     )
-    assert transition.start_slot == offset_slots[0]
-
+    # Process the shard transition; in the mean time, reconstruct the headers for signature
+    # verification and the proposers for reward/penalty calculation
     headers = []
     proposers = []
-    prev_gasprice = state.shard_states[shard].gasprice
-    shard_parent_root = state.shard_states[shard].latest_block_root
-    for i, offset_slot in enumerate(offset_slots):
+    for i, slot in enumerate(admissible_slots):
         shard_block_length = transition.shard_block_lengths[i]
         shard_state = transition.shard_states[i]
         # Verify correct calculation of gas prices and slots
         assert shard_state.gasprice == compute_updated_gasprice(prev_gasprice, shard_block_length)
-        assert shard_state.slot == offset_slot
+        assert shard_state.slot == slot
         # Collect the non-empty proposals result
-        is_empty_proposal = shard_block_length == 0
-        if not is_empty_proposal:
-            proposal_index = get_shard_proposer_index(state, offset_slot, shard)
+        if shard_block_length > 0:
+            shard_proposer_index = get_shard_proposer_index(state, slot, shard)
             # Reconstruct shard headers
             header = ShardBlockHeader(
                 shard_parent_root=shard_parent_root,
-                beacon_parent_root=get_block_root_at_slot(state, offset_slot),
-                slot=offset_slot,
+                beacon_parent_root=get_block_root_at_slot(state, slot),
+                slot=slot,
                 shard=shard,
-                proposer_index=proposal_index,
+                proposer_index=shard_proposer_index,
                 body_root=transition.shard_data_roots[i]
             )
             shard_parent_root = hash_tree_root(header)
```

Comment:
Do we want to validate `shard_parent_root` here? Feel free to add that check in this PR so i can close #2006

Reply:
We don't need to "validate" it; we're rather _reconstructing_ it. That is, we first pull the parent root from the most recent block hash in the shard state to generate the first header. We then use the put the root of the first header as the parent root of the second header, etc etc.

Reply:
Sorry, i meant more on verifying the constructed header is correct by comparing it to shard state's latest block root.

```
header: {slot:100.... } // constructed
shard_state: {slot:100, latest_block_root: 0xA} // from transition
assert hash_tree_root(header) == shard_state.latest_block_root // verify header root is 0xA
```